Source/Bogus/Extensions/Norway/ExtensionsForNorway.cs

```diff
@@ -0,0 +1,119 @@
+﻿using System;
+
+namespace Bogus.Extensions.Norway
+{
+   /// <summary>
+   /// API extensions specific for a geographical location.
+   /// </summary>
+   public static class ExtensionsForNorway
+   {
+      /// <summary>
+      /// Norwegian national identity number (fødselsnummer)
+      /// </summary>
+      public static string Fødselsnummer(this Person p)
+      {
+         const string Key = nameof(ExtensionsForNorway) + "Fødselsnummer";
+         if (p.context.ContainsKey(Key))
+         {
+            return p.context[Key] as string;
+         }
+
+         /*
+             DDMMYYXXXCC
+             | | | |  |--> Checksum
+             | | | |
+             | | | |
+             | | | |-----> Individual number
+             | | |-------> Year (last two digits)
+             | |---------> Month
+             |-----------> Day
+
+             The individual number has to be even for women and odd for men.
+
+             The checksum is calculated with a modulo checksum algorithm.
+             If either of the checksum numbers are 10, the fødselsnummer gets
+             rejected, and a new individual number has to be generated.
+
+            https://www.skatteetaten.no/en/person/national-registry/birth-and-name-selection/children-born-in-norway/national-id-number/
+         */
+
+         var r = p.Random;
+         string birthDate = $"{p.DateOfBirth:ddMMyy}";
+
+         string individualNumber;
+         string checksum;
+         bool isOkChecksum;
+
+         do
+         {
+            individualNumber = GenerateIndividualNumber(r, p.Gender, p.DateOfBirth.Year);
+            isOkChecksum = GenerateChecksum(birthDate, individualNumber, out checksum);
+         } while (!isOkChecksum);
```

Comment:
Hi @bchavez,

The reason for the loop is because the checksums can end up getting 10 as calculated value , either cs1, cs2 or both. In that case the number is invalid, and a new one has to be generated. The only thing that can be regenerated is the individual number, because the birthdate is fixed. So in the loop I generate a new individual number and try calculating the checksum again.

I don't think it will break determinism, as long as the global Random object is used. Given a seed the loop will run the same number of iterations every time.

As for the second comment:

I am not sure why the weights are as they are in the Python library, but they should be correct in the PR. The offical spec is [here](https://git.sarepta.ehelse.no/publisert/standarder/raw/master/kravdokument/IdentifikatorerForPersoner/HIS_1001_2010_Identifikatorer%20for%20personer%20v1.pdf) (page 12-13):

In English:
> 
> The weights are:
> 3, 7, 6, 1, 8, 9, 4, 5 and 2
> k1 = 11 - ((3*D1 + 7*D2 + 6*M1 + 1*M2 + 8*Y1 + 9*Y2 + 4*I1 + 5*I2 + 2*I3) - 11*Q1)
> 
> Q1 is the the integer quotient for the product sum divided by 11, meaning the highest number that when multiplied with 11 gives the product sum or less.
> 
> [Q1 is just modulo.]
> 
> There are no birth numbers assigned that gives k1 = 10.
> If k1 = 11, set K1 equal to 0, otherwise K1 is equal k1.
> 
> The second control digits, K2, is calculated in the same way of the first 9 digits and the first
> control digit.
> 
> The weights here are:
> 5, 4, 3, 2, 7, 6, 5, 4, 3 and 2
> k2 = 11 - ((5*D1 + 4*D2 + 3*M1 + 2*M2 + 7*Y1 + 6*Y2 + 5*I1 + 4*I2 + 3*I3 + 2*K1) - 11*Q2)
> 
> Q2 is the the integer quotient for the product sum divided by 11, meaning the highest number that when multiplied with 11 gives the product sum or less.
> 
> There are no birth numbers assigned that gives k2 = 10.
> If k2 = 11, set K2 equal to 0, otherwise K2 is equal k2.


I generated a couple of numbers and double checked their validity with a couple of other tools, e.g. [this](http://www.fnrinfo.no/Verktoy/Kontroll.aspx) and [this](http://www.lefdal.cc/div/mod11-sjekk.php) (first textbox), and they said the numbers were valid.

These are the numbers I checked (base64 encoded to make sure Google doesn't index numbers that belong to real people):

MjQwMzk3MDQxNDENCjEwMDE4MTI5MDIyDQowNDExOTkwMjQzMQ0KMTgwMTYyMDQxNjgNCjEzMDM4NzM1NzQy

The calculation of the checksum could of course be made in a different way. A functional approach would be to use map-fold. Another imperative approach is to use a loop that iterates over a split birthdate+indnum string and then multiplies each number with the weight and adds it to a total variable (like in the Python code).

As for the do-while loop, another approach there could be to increment or decrement the individual number if the control digits become 10, and then recalculate the control digits with the incremented/decremented values. The control digits should then become something else than 10, but I cannot guarantee it. The incremented/decremented individual number would become the new individual number of course, it's not just done in the checksum calculation.